### PR TITLE
TFKUBE-369: Maintain stickiness by always provisioning an ingress

### PR DESCRIPTION
**Pull request by errcode1202**
- [x] [new e2e tests with no domain passing](https://github.com/atlassian-labs/data-center-terraform/runs/5534920838?check_suite_focus=true)
Output for `ingress` and `product_urls` for this test below:
```shell
ingress = {
  "certificate" = "Certificate is not provisioned"
  "load_balancer_hostname" = "a910e28b9ccc24765aabb981e3f1f039-1015236823.eu-central-1.elb.amazonaws.com"
}
product_urls = {
  "bamboo" = "http://a910e28b9ccc24765aabb981e3f1f039-1015236823.eu-central-1.elb.amazonaws.com/bamboo"
  "bitbucket" = "http://a910e28b9ccc24765aabb981e3f1f039-1015236823.eu-central-1.elb.amazonaws.com/bitbucket"
  "confluence" = "http://a910e28b9ccc24765aabb981e3f1f039-1015236823.eu-central-1.elb.amazonaws.com/confluence"
  "jira" = "http://a910e28b9ccc24765aabb981e3f1f039-1015236823.eu-central-1.elb.amazonaws.com/jira"
}
```

- [x] [new e2e tests with domain passing ](https://github.com/atlassian-labs/data-center-terraform/runs/5533776490?check_suite_focus=true) 
Output for `ingress` and `product_urls` for this test below:
```shell
ingress = {
  "certificate" = "arn:aws:acm:ap-south-1:887764444972:certificate/83beacc2-3f8b-4297-9274-85b2795e8484"
  "load_balancer_hostname" = "ad606e47e1f7a4e25b1675c5a8ec28f3-2051422123.ap-south-1.elb.amazonaws.com"
}
product_urls = {
  "bamboo" = "https://bamboo.e2etest-t7f0vd.deplops.com/"
  "bitbucket" = "https://bitbucket.e2etest-t7f0vd.deplops.com/"
  "confluence" = "https://confluence.e2etest-t7f0vd.deplops.com/"
  "jira" = "https://jira.e2etest-t7f0vd.deplops.com/"
}
```
- [x] [current e2e tests passing](https://github.com/atlassian-labs/data-center-terraform/runs/5524579580?check_suite_focus=true)
- [x] stickiness works for both scenarios (with `domain` and without)
- [x] bitbucket ssh maintained for both scenarios (with `domain` and without)
- [x] scaling product instance count works for both scenarios (with `domain` and without)
- [x] unit tests

### CODE REVIEW

File: modules/products/bamboo/locals.tf [hunk 1]
```diff
@@ -28,29 +28,38 @@ locals {
...
+        }
+      ]
+    }
+  }) : yamlencode({})
+
+  context_path_settings = !local.domain_supplied ? yamlencode({
     bamboo = {
       service = {
-        type = "LoadBalancer"
+        contextPath = "/bamboo"
```
Comment (errcode1202) [inline modules/products/bamboo/locals.tf:59]: update

File: modules/products/bamboo/locals.tf [hunk 2]
```diff
@@ -28,29 +28,38 @@ locals {
...
+
   # ingress settings for Bamboo service
-  ingress_with_domain = yamlencode({
+  ingress_settings = yamlencode({
     ingress = {
       create = "true"
-      host   = local.product_domain_name
+      host   = local.domain_supplied ? "${local.product_name}.${var.ingress.outputs.domain}" : var.ingress.outputs.lb_hostname
+      https  = local.domain_supplied ? true : false
+      path   = local.domain_supplied ? null : "/bamboo"
```
Comment (errcode1202) [inline modules/products/bamboo/locals.tf:40]: update

File: modules/products/bamboo/outputs.tf [hunk 3]
```diff
@@ -1,5 +1,5 @@
 output "product_domain_name" {
-  value = local.use_domain ? "https://${local.product_domain_name}" : "http://${data.kubernetes_service.bamboo.status[0].load_balancer[0].ingress[0].hostname}"
+  value = local.domain_supplied ? "https://${local.product_domain_name}" : "http://${var.ingress.outputs.lb_hostname}/bamboo"
```
Comment (errcode1202) [inline modules/products/bamboo/outputs.tf:2]: update with `local.product_name`

**Review by nanux — APPROVED** [review]
Reviewed most of the code when paired with @errcode1202.

File: modules/AWS/ingress/main.tf [hunk 4]
```diff
@@ -91,6 +107,8 @@ resource "helm_release" "ingress" {
     # reflected in the associated ELB Load Balancer. As such, the method "enable_ssh_tcp_protocol_on_lb_listener" (install.sh)
     # is executed, post deployment, to update the protocol on the load balancer from HTTP to TCP.
     #
+    local.aws_load_balancer_ssl_cert,
```
[the remark this answers is not in the document]
Comment (errcode1202) [reply modules/AWS/ingress/main.tf:110]: Comments for each are in the `local.tf` for the ingress module:

- https://github.com/atlassian-labs/data-center-terraform/blob/5b395dbedfbe537a585c4b2328b6feb515661cd2/modules/AWS/ingress/locals.tf#L13
- https://github.com/atlassian-labs/data-center-terraform/blob/5b395dbedfbe537a585c4b2328b6feb515661cd2/modules/AWS/ingress/locals.tf#L25

File: modules/AWS/ingress/outputs.tf [hunk 5]
```diff
@@ -1,8 +1,8 @@
-output "ingress" {
+output "outputs" {
```
[the remark this answers is not in the document]
Comment (errcode1202) [reply modules/AWS/ingress/outputs.tf:1]: Let's re-think this as part of the work for: https://bulldog.internal.atlassian.com/browse/TFKUBE-388

**Pull request merged**